docs(mnnvl): update GREP-417 to single-annotation model (#568)
ai-dynamo/grove: docs/proposals/417-auto-mnnvl/README.md

## Summary

Grove's auto-MNNVL feature allows users to leverage NVIDIA Multi-Node NVLink acceleration through a simple annotation — without manually authoring `ComputeDomain` resources or wiring `resourceClaims` in their pod specs. The operator automatically creates and manages ComputeDomains, injects RCT references, and handles scaling as PCS replicas scale up or down.

The feature is controlled by a single annotation:

> - `grove.io/mnnvl-group`: a string name — assigns the PodClique to a named MNNVL group (ComputeDomain). The special value `"none"` explicitly opts out of MNNVL, overriding any parent-level setting.

Reviewer: case-sensitive or not ?

Author: Case-sensitive, following the Kubernetes convention for annotation values. Only the exact lowercase `"none"` is recognized as opt-out. Added a note to the doc to clarify this.

> - `grove.io/mnnvl-group`: a string name — assigns the PodClique to a named MNNVL group (ComputeDomain). The special value `"none"` (case-sensitive) explicitly opts out of MNNVL, overriding any parent-level setting.

Reviewer 2: ```suggestion
- `grove.io/mnnvl-group`: <group name> — assigns a Grove workload to a named MNNVL group (ComputeDomain). The user may choose any group name as the annotation value except the reserved strings: `"none"` and `"disabled"`. These reserved special values `"none"` and `"disabled"` explicitly opts out of MNNVL, overriding any parent-level setting.
```

Reviewer 3: how about "a Grove primitive" or "A Grove CR" or something like that, I think "a Grove workload" doesn't properly express that its at various levels - the part below explains it but I think workload makes it seem just like you can set it at PCS to me.

> The annotation can be placed on a **PodCliqueSet**, **PodCliqueScalingGroup**, or **PodClique**, and propagates downward (PCS → PCSG → PCLQ) with lower layers overriding higher ones.

Reviewer 3: ```suggestion
The annotation can be placed on a **PodCliqueSet**, **PodCliqueScalingGroup**, or **PodClique**, and propagates downward (PCS → PCSG → PCLQ) with values on lower layers overriding higher ones.
```

Author: Good catch, accepted. Applied in the next push.

- **Granularity:** Users can control MNNVL at the PodClique level — only the pods that need NVLink are enrolled, others are unaffected.
- **Heterogeneous cluster support:** In clusters where only some nodes have NVIDIA IMEX driver support and some nodes don't, users can scope MNNVL to specific PodCliques, avoiding failures on unsupported nodes.
- **Composability:** Multiple MNNVL groups within a single PCS replica allow different PodCliques to participate in separate ComputeDomains.

### Goals

- Provide a **single annotation** (`grove.io/mnnvl-group`) that controls both MNNVL enrollment and group assignment — zero conflict states, one annotation to manage.
- Support **multi-layer propagation** — the annotation can be placed on PCS, PCSG, or PCLQ, with lower layers overriding higher ones.
- Use an **opt-in model** — no MNNVL is applied unless the annotation is present.
- Have the operator **automatically manage ComputeDomain lifecycle** (create, scale, delete, protect) per group per PCS replica.
- Support **heterogeneous clusters** — only annotated PodCliques need NVIDIA DRA-capable nodes.
- Support **multiple MNNVL groups** within a single PCS replica.

> - Remove/Deprecate the Phase 1 `grove.io/auto-mnnvl` annotation — not backward compatible (acceptable in alpha).

Reviewer 3: ```suggestion
- Remove the Phase 1 `grove.io/auto-mnnvl` annotation — not backward compatible (acceptable in alpha).
```

Author: Agreed, accepted. Applied in the next push.

metadata:
  name: training-job
  annotations:
    grove.io/mnnvl-group: "my-group"    # MNNVL for all GPU PCLQs
spec:
  template:
    cliques:
      - name: workers              # inherits "my-group" from PCS
        spec: ...
      - name: param-servers
        annotations:
          grove.io/mnnvl-group: "none"  # override: no MNNVL for this PCLQ
        spec: ...
```

The operator discovers this annotation, determines the set of ComputeDomains required per replica, and:
- manages their full lifecycle — creation, deletion, scaling, and protection via finalizers.
- injects RCT references into the pod's spec of enrolled PodCliques.

**Key behavioral rules:**
- **Opt-in:** No MNNVL is applied unless `mnnvl-group` is present at some layer. Absence of the annotation at all layers means no MNNVL.
- **Propagation:** The annotation propagates downward (PCS → PCSG → PCLQ). A lower layer overrides a higher layer.

> - **Opt-out:** Setting `mnnvl-group: "none"` explicitly opts out of MNNVL, overriding any parent-level setting. The `"none"` value is reserved and cannot be used as a group name.

Reviewer 2: ```suggestion
- **Opt-out:** Setting `mnnvl-group: "none" | "disabled"` explicitly opts out of MNNVL, overriding any parent-level setting. The `"none"` value is reserved and cannot be used as a group name.
```

### Annotation Semantics

MNNVL participation is controlled by a single annotation:

#### `grove.io/mnnvl-group` — MNNVL group assignment

```yaml
grove.io/mnnvl-group: "<value>"
```

| Value | Meaning |
|---|---|
| String name (`"default"`, `"training"`, `"workers"`, ...) | Assign the PCLQ to a named MNNVL group. PodCliques with the same group name share a ComputeDomain per replica. |
| `"none"` | Explicit opt-out. Used to override a parent layer's group assignment. Reserved — cannot be used as a group name. |

> | Absent | Inherit from parent layer. If no parent has it, no MNNVL. |

Reviewer 3: Do we need Absent as an option? I thought in this case the user just wouldn't put the annotation instead of saying Absent.

Author: "Absent" isn't a value the user writes — it documents the operator's behavior when the annotation is not present on a given layer. It's needed because the operator distinguishes between three states: a group name (enrolled), `"none"` (explicit opt-out), and not present (inherit from parent). Without the Absent row, the inheritance/propagation behavior would be ambiguous — e.g., it wouldn't be clear whether a missing annotation means "no MNNVL" or "inherit from parent".

> **Group name validation:** Values are **case-sensitive**, following the Kubernetes convention for annotation values. The value must be either `"none"` (opt-out) or a valid Kubernetes name component: lowercase alphanumeric characters or dashes, must start and end with an alphanumeric character, max 63 characters. This is required because the group name becomes part of the ComputeDomain resource name. Empty strings and values containing invalid characters are rejected.

Reviewer 2: ```suggestion
**Group name validation:** Values are **case-sensitive**, following the Kubernetes convention for annotation values. The value must be a valid Kubernetes name component: lowercase alphanumeric characters or dashes, must start and end with an alphanumeric character, max 63 characters. This is required because the group name becomes part of the ComputeDomain resource name. Reserved strings `"none"` and `"disabled"` are used to opt-out of the auto-mnnvl feature. Empty strings and values containing invalid characters are rejected.
```

> **Non-GPU PCLQs:** Non-GPU PCLQs that resolve to MNNVL enrollment (whether explicit or inherited) are silently skipped — no RCT injection, no error. This allows PCS-level defaults without requiring `"none"` overrides on every non-GPU PCLQ.

Reviewer 2: ```suggestion
**Non-GPU PCLQs:** Non-GPU PCLQs that resolve to MNNVL enrollment (whether explicit or inherited) are silently skipped — no RCT injection, no error. This allows PCS-level defaults without requiring `"none"` or `"disabled"` overrides on every non-GPU PCLQ.
```

3. **RCT reference injection:** Injected into the PCLQ's pod spec template at **PCLQ creation time** (not at pod creation time). This early-binding approach ensures the decision is made once and baked into the PCLQ spec. The injection flow depends on whether the PCLQ is standalone or managed by a PCSG:
   - **PCS → standalone PCLQ:** For PCLQs defined directly in the PCS template (not managed by a PCSG), the PCS controller resolves the effective MNNVL state (from `mnnvl-group` annotations at PCS vs. PCLQ level) and injects `resourceClaims` if enrolled.
   - **PCS → PCSG:** The PCS controller propagates the `grove.io/mnnvl-group` annotation to the PCSG (if the PCSG doesn't already have its own value defined in the PCS template).
   - **PCSG → PCLQ:** For PCLQs managed by a PCSG, the PCSG controller resolves the effective MNNVL state (PCSG annotation vs. PCLQ template annotation) and injects the RCT reference.
   - **Pod creation:** No special logic — pods use the PCLQ's pod spec as-is.

4. **Non-enrolled PodCliques:** Left untouched — no CD reference injected.

#### CD Naming Convention

> CD names are included at group name: `{pcs-name}-{replica-index}-{group-name}`.

Reviewer 4: ```suggestion
CD names include the group name and use the form `{pcs-name}-{replica-index}-{group-name}`.
```

    FeatureOff --> PCSSubmitOff["PCS Submitted"]
    FeatureOn --> PCSSubmitOn["PCS Submitted"]

    PCSSubmitOff --> HasAnnotationOff{"Has mnnvl-group\nannotation?"}
    HasAnnotationOff -->|No| AcceptNoMNNVL1["Accept: no MNNVL"]
    HasAnnotationOff -->|Yes| RejectDisabled["Reject: feature disabled"]

    PCSSubmitOn --> ValidateValues{"Valid mnnvl-group\nvalues?"}
    ValidateValues -->|No| RejectInvalid["Reject: invalid\nannotation value"]
    ValidateValues -->|Yes| ResolveGroups["Resolve MNNVL enrollment\nfor each PCLQ"]
    ResolveGroups --> AnyEnrolled{"Any PCLQ enrolled?"}

    AnyEnrolled -->|No| AcceptNoMNNVL2["Accept: no MNNVL"]

>     AnyEnrolled -->|Yes| CreateCDs["Create CDs per group\nInject RCT for GPU PCLQs\nNon-GPU PCLQs silently skipped"]

Reviewer 4: The flowchart's Any PCLQ enrolled? → Create CDs per group branch doesn't show the non-GPU filter, but the lifecycle section (line 321) says group collection ignores non-GPU PCLQs?

```

### Webhook Behavior

#### Mutating Webhook (on Create)

In the opt-in model, the mutating webhook **does not** add any MNNVL-related annotations automatically. The user is responsible for providing `grove.io/mnnvl-group`.

#### Validating Webhook (on Create)

The validating webhook enforces:

> - **Value validation (`mnnvl-group`):** Must be `"none"` (opt-out) or a valid Kubernetes name component (lowercase alphanumeric or dashes, starting and ending with alphanumeric, max 63 chars). Invalid values are rejected.

Reviewer 2: ```suggestion
- **Value validation (`mnnvl-group`):** Must be `"none"` or `"disabled"` (opt-out) or a valid Kubernetes name component (lowercase alphanumeric or dashes, starting and ending with alphanumeric, max 63 chars). Invalid values are rejected.
```

### Test Plan

> - **Unit tests:** Cover annotation resolution (propagation, override, `"none"` opt-out), group name validation (valid K8s name component + `"none"` reserved), ComputeDomain lifecycle management (create/delete/scale), RCT injection logic, CD naming (`{pcs}-{replica}-{group}`), and configuration parsing.

Reviewer 2: ```suggestion
- **Unit tests:** Cover annotation resolution (propagation, override, `"none"` or `"disabled"` opt-out), group name validation (valid K8s name component + `"none"` reserved), ComputeDomain lifecycle management (create/delete/scale), RCT injection logic, CD naming (`{pcs}-{replica}-{group}`), and configuration parsing.
```

- **E2E tests:**
  - PCS with `mnnvl-group: "my-group"` → single CD per replica, all GPU PCLQs enrolled.
  - PCS with `mnnvl-group` on some PCLQs → per-group CDs, only enrolled GPU PCLQs get RCT.
  - PCS with PCS-level `mnnvl-group: "my-group"` + PCLQ `mnnvl-group: "none"` → PCLQ excluded.
  - PCS with multiple group names → separate CDs per group per replica.
  - PCS with `mnnvl-group` on non-GPU PCLQ → silently skipped, no RCT for that PCLQ.
  - PCS scale-up/scale-down → CDs created/deleted accordingly.
  - `autoMNNVLEnabled: false` + annotations → rejected.
  - `autoMNNVLEnabled: true` + CRD absent → operator fails to start.
  - Annotation immutability: update attempts rejected.

### Graduation Criteria

#### Alpha
- `grove.io/mnnvl-group` annotation implemented and functional.
- Annotation propagation (PCS → PCSG → standalone PCLQ) working.
- CD naming (`{pcs}-{replica}-{group}`) working.

> - `"none"` opt-out enforced.

Reviewer 2: ```suggestion
- `"none"` or `"disabled"` opt-out enforced.
```

- Boolean configuration (`autoMNNVLEnabled`) unchanged from Phase 1 — no migration required.
- Unit and basic integration tests passing.

#### Beta

> - E2E tests covering all key scenarios from the behavior matrix.

Reviewer 4: Nit: This still references the behavior matrix, but this PR removes that section.
